[GStreamer][MediaStream] video capture source for Camera portal

philn opened the pull request:
#### c28c004edc00e3010cb74e7aba5b17d80bd2f99e
<pre>
[GStreamer][MediaStream] video capture source for Camera portal
<a href="https://bugs.webkit.org/show_bug.cgi?id=244004">https://bugs.webkit.org/show_bug.cgi?id=244004</a>

Reviewed by Xabier Rodriguez-Calvar.

When available on a desktop host, WebKitGTK can now leverage the XDG desktop portal for Camera
access so no specific sandbox exception is required. After successful communication through the
portal, the WebProcess receives a FD pointing to the PipeWire session created by the portal. To
expose the device we then create a GStreamer pipewiredeviceprovider and pass the fd to it.

* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr&lt;GstDeviceProvider&gt;):
(WTF::derefGPtr&lt;GstDeviceProvider&gt;):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp: Added.
(WebCore::createDBusProxy):
(WebCore::DesktopPortalCamera::create):
(WebCore::DesktopPortalCamera::DesktopPortalCamera):
(WebCore::DesktopPortalScreenCast::create):
(WebCore::DesktopPortalScreenCast::DesktopPortalScreenCast):
(WebCore::DesktopPortal::DesktopPortal):
(WebCore::DesktopPortal::getProperty):
(WebCore::DesktopPortal::waitResponseSignal):
(WebCore::DesktopPortalCamera::isCameraPresent):
(WebCore::DesktopPortalCamera::accessCamera):
(WebCore::DesktopPortalCamera::openCameraPipewireRemote):
(WebCore::DesktopPortalScreenCast::createScreencastSession):
(WebCore::DesktopPortalScreenCast::closeSession):
(WebCore::DesktopPortalScreenCast::ScreencastSession::selectSources):
(WebCore::DesktopPortalScreenCast::ScreencastSession::start):
(WebCore::DesktopPortalScreenCast::ScreencastSession::openPipewireRemote):
* Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.h: Added.
(WebCore::DesktopPortal::waitResponseSignal):
(WebCore::DesktopPortal::notifyResponse):
(WebCore::DesktopPortalScreenCast::ScreencastSession::ScreencastSession):
(WebCore::DesktopPortalScreenCast::ScreencastSession::path const):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.cpp:
(WebCore::GStreamerAudioCapturer::GStreamerAudioCapturer):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::addDevice):
(WebCore::GStreamerCaptureDeviceManager::refreshCaptureDevices):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.cpp:
(WebCore::GStreamerCapturer::GStreamerCapturer):
(WebCore::GStreamerCapturer::createSource):
(WebCore::GStreamerCapturer::caps):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerDisplayCaptureDeviceManager.cpp:
(WebCore::GStreamerDisplayCaptureDeviceManager::createDisplayCaptureSource):
(WebCore::GStreamerDisplayCaptureDeviceManager::stopSource):
(WebCore::GStreamerDisplayCaptureDeviceManager::waitResponseSignal): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.cpp:
(WebCore::GStreamerVideoCaptureDeviceManager::GStreamerVideoCaptureDeviceManager):
(WebCore::GStreamerVideoCaptureDeviceManager::computeCaptureDevices):
(WebCore::GStreamerVideoCaptureDeviceManager::createVideoCaptureSource):
(WebCore::GStreamerVideoCaptureSource::createPipewireSource):
(WebCore::GStreamerVideoCaptureSource::GStreamerVideoCaptureSource):
(WebCore::m_capturer):
(WebCore::GStreamerVideoCaptureSource::settingsDidChange):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
(WebCore::GStreamerVideoCapturer::GStreamerVideoCapturer):
(WebCore::GStreamerVideoCapturer::isCapturingDisplay const):
(WebCore::GStreamerVideoCapturer::setupPipeline):
(WebCore::GStreamerVideoCapturer::setSize):
(WebCore::GStreamerVideoCapturer::reconfigure):
(WebCore::GStreamerVideoCapturer::createSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDevice.h: Copied from Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h.
(WebCore::PipeWireCaptureDevice::PipeWireCaptureDevice):
(WebCore::PipeWireCaptureDevice::objectId const):
(WebCore::PipeWireCaptureDevice::fd const):
(WebCore::PipeWireCaptureDevice::caps const):
* Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDeviceManager.cpp: Added.
(WebCore::PipeWireCaptureDeviceManager::create):
(WebCore::PipeWireCaptureDeviceManager::PipeWireCaptureDeviceManager):
(WebCore::PipeWireCaptureDeviceManager::computeCaptureDevices):
(WebCore::PipeWireCaptureDeviceManager::createCaptureSource):
* Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDeviceManager.h: Added.
(WebCore::PipeWireCaptureDeviceManager::captureDevices const):
* Source/WebCore/platform/mediastream/gstreamer/PipeWireNodeData.h: Copied from Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h.
(WebCore::PipeWireNodeData::PipeWireNodeData):
* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bubblewrapSpawn):

Canonical link: <a href="https://commits.webkit.org/292932@main">https://commits.webkit.org/292932@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c8633a4908a9827ca01b529ae5879d558a68748

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97503 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17128 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102592 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48032 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99548 "Build is in progress. Recent messages:OS: Sequoia (15.3), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed bindings tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74298 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31477 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100506 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13196 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/88178 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54643 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12965 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6059 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47474 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82994 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6140 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104610 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24582 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17933 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83343 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24954 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84303 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82764 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20822 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27309 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4997 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18176 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24544 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29713 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24366 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27680 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

csaavedra commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp:
```diff
@@ -0,0 +1,311 @@
+
+static const Seconds s_dbusCallTimeout = 10_ms;
+
+GRefPtr<GDBusProxy> createDBusProxy(const String& interfaceName)
```
As far as I can see there's no reason for this to be a `String`, you can use an `ASCIILIteral` for less overhead.

calvaris replied:
I agree and then you can use `ASCIILiteral::characters()` that will give you the null terminated string inside it.

csaavedra commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.h:
```diff
@@ -0,0 +1,98 @@
+    void notifyResponse(GVariant* parameters) { m_currentResponseCallback(parameters); }
+
+protected:
+    String m_interfaceName;
```
This can also be an `ASCIILiteral`.

calvaris replied:
I agree with Claudio here. And you can change the method calls all the way through the onion.

csaavedra commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp:
```diff
@@ -0,0 +1,311 @@
+/*
+ * Copyright (C) 2025 Igalia S.L
```
This and other new headers are missing the dot after L.

csaavedra reviewed (commented):
A few comments.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp:
```diff
@@ -0,0 +1,311 @@
+{
+}
+
+GRefPtr<GVariant> DesktopPortal::getProperty(const char* name)
```
I would use ASCIILiteral here as well.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp:
```diff
@@ -0,0 +1,311 @@
+    return nullptr;
+}
+
+void DesktopPortal::waitResponseSignal(const char* objectPath, ResponseCallback&& callback)
```
ASCIILiteral?

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp:
```diff
@@ -0,0 +1,311 @@
+    g_variant_get(result.get(), "(o)", &objectPath.outPtr());
+    waitResponseSignal(objectPath.get());
+
+    auto requestPath = String::fromLatin1(objectPath.get());
```
I think you can create a StringView here.

calvaris commented on Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDeviceManager.cpp:
```diff
@@ -0,0 +1,109 @@
+    if (MockRealtimeMediaSourceCenter::mockRealtimeMediaSourceCenterEnabled())
+        return;
+
+    if (!m_pipewireDeviceProvider || !gstObjectHasProperty(GST_OBJECT_CAST(m_pipewireDeviceProvider.get()), "fd")) {
```
Isn't `gstObjectHasProperty` requiring an ASCIILiteral already? Wouldn't you need `"fd"_s` here?

csaavedra commented on Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp:
```diff
@@ -0,0 +1,311 @@
+    return adoptRef(*new DesktopPortalCamera(interfaceName, WTFMove(proxy)));
+}
+
+DesktopPortalCamera::DesktopPortalCamera(const String& interfaceName, GRefPtr<GDBusProxy>&& proxy)
```
This `String` slip in.

philn committed:
[GStreamer][MediaStream] video capture source for Camera portal
https://bugs.webkit.org/show_bug.cgi?id=244004

Reviewed by Xabier Rodriguez-Calvar.

When available on a desktop host, WebKitGTK can now leverage the XDG desktop portal for Camera
access so no specific sandbox exception is required. After successful communication through the
portal, the WebProcess receives a FD pointing to the PipeWire session created by the portal. To
expose the device we then create a GStreamer pipewiredeviceprovider and pass the fd to it.

* Source/WebCore/platform/SourcesGStreamer.txt:
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.cpp:
(WTF::adoptGRef):
(WTF::refGPtr<GstDeviceProvider>):
(WTF::derefGPtr<GstDeviceProvider>):
* Source/WebCore/platform/graphics/gstreamer/GRefPtrGStreamer.h:
* Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.cpp: Added.
(WebCore::createDBusProxy):
(WebCore::DesktopPortalCamera::create):
(WebCore::DesktopPortalCamera::DesktopPortalCamera):
(WebCore::DesktopPortalScreenCast::create):
(WebCore::DesktopPortalScreenCast::DesktopPortalScreenCast):
(WebCore::DesktopPortal::DesktopPortal):
(WebCore::DesktopPortal::getProperty):
(WebCore::DesktopPortal::waitResponseSignal):
(WebCore::DesktopPortalCamera::isCameraPresent):
(WebCore::DesktopPortalCamera::accessCamera):
(WebCore::DesktopPortalCamera::openCameraPipewireRemote):
(WebCore::DesktopPortalScreenCast::createScreencastSession):
(WebCore::DesktopPortalScreenCast::closeSession):
(WebCore::DesktopPortalScreenCast::ScreencastSession::selectSources):
(WebCore::DesktopPortalScreenCast::ScreencastSession::start):
(WebCore::DesktopPortalScreenCast::ScreencastSession::openPipewireRemote):
* Source/WebCore/platform/mediastream/gstreamer/DesktopPortal.h: Added.
(WebCore::DesktopPortal::waitResponseSignal):
(WebCore::DesktopPortal::notifyResponse):
(WebCore::DesktopPortalScreenCast::ScreencastSession::ScreencastSession):
(WebCore::DesktopPortalScreenCast::ScreencastSession::path const):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.cpp:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.cpp:
(WebCore::GStreamerAudioCapturer::GStreamerAudioCapturer):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.cpp:
(WebCore::GStreamerCaptureDeviceManager::addDevice):
(WebCore::GStreamerCaptureDeviceManager::refreshCaptureDevices):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCaptureDeviceManager.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.cpp:
(WebCore::GStreamerCapturer::GStreamerCapturer):
(WebCore::GStreamerCapturer::createSource):
(WebCore::GStreamerCapturer::caps):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerDisplayCaptureDeviceManager.cpp:
(WebCore::GStreamerDisplayCaptureDeviceManager::createDisplayCaptureSource):
(WebCore::GStreamerDisplayCaptureDeviceManager::stopSource):
(WebCore::GStreamerDisplayCaptureDeviceManager::waitResponseSignal): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.cpp:
(WebCore::GStreamerVideoCaptureDeviceManager::GStreamerVideoCaptureDeviceManager):
(WebCore::GStreamerVideoCaptureDeviceManager::computeCaptureDevices):
(WebCore::GStreamerVideoCaptureDeviceManager::createVideoCaptureSource):
(WebCore::GStreamerVideoCaptureSource::createPipewireSource):
(WebCore::GStreamerVideoCaptureSource::GStreamerVideoCaptureSource):
(WebCore::m_capturer):
(WebCore::GStreamerVideoCaptureSource::settingsDidChange):
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.cpp:
(WebCore::GStreamerVideoCapturer::GStreamerVideoCapturer):
(WebCore::GStreamerVideoCapturer::isCapturingDisplay const):
(WebCore::GStreamerVideoCapturer::setupPipeline):
(WebCore::GStreamerVideoCapturer::setSize):
(WebCore::GStreamerVideoCapturer::reconfigure):
(WebCore::GStreamerVideoCapturer::createSource): Deleted.
* Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoCapturer.h:
* Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDevice.h: Copied from Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h.
(WebCore::PipeWireCaptureDevice::PipeWireCaptureDevice):
(WebCore::PipeWireCaptureDevice::objectId const):
(WebCore::PipeWireCaptureDevice::fd const):
(WebCore::PipeWireCaptureDevice::caps const):
* Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDeviceManager.cpp: Added.
(WebCore::PipeWireCaptureDeviceManager::create):
(WebCore::PipeWireCaptureDeviceManager::PipeWireCaptureDeviceManager):
(WebCore::PipeWireCaptureDeviceManager::computeCaptureDevices):
(WebCore::PipeWireCaptureDeviceManager::createCaptureSource):
* Source/WebCore/platform/mediastream/gstreamer/PipeWireCaptureDeviceManager.h: Added.
(WebCore::PipeWireCaptureDeviceManager::captureDevices const):
* Source/WebCore/platform/mediastream/gstreamer/PipeWireNodeData.h: Copied from Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCapturer.h.
(WebCore::PipeWireNodeData::PipeWireNodeData):
* Source/WebKit/UIProcess/Launcher/glib/BubblewrapLauncher.cpp:
(WebKit::bubblewrapSpawn):

Canonical link: https://commits.webkit.org/292932@main
